geopandas/clip.py

```diff
@@ -0,0 +1,245 @@
+    sidx = list(spatial_index.intersection(bbox))
+    gdf_sub = gdf.iloc[sidx]
+
+    # Clip the data - with these data
+    clipped = gdf_sub.copy()
+    clipped["geometry"] = gdf_sub.intersection(poly)
+
+    # Return the clipped layer with no null geometry values or empty geometries
+    return clipped[~clipped.geometry.is_empty & clipped.geometry.notnull()]
+
+
+def _clip_multi_poly_line(gdf, poly):
+    """Clip multi lines and polygons to the clip_obj GeoDataFrame extent.
+
+    Clip an input multi line or polygon to the polygon extent of the clip_obj
+    parameter. Lines or Polygons that intersect the clip_obj geometry are
+    extracted with associated attributes and returned.
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+        multiLine or multipolygon geometry that is clipped to clip_obj.
+
+    poly : (Multi)Polygon
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of gdf
+        that intersects with clip_obj.
+    """
+
+    # Clip multi polygons
+    # Explode multi polygons so that intersection works with all parts of the polygon
+    # If this step isn't taken, intersection only gets the intersection of one part of
+    # the multi part object. Also reset index to remove the column made by explode.
+    clipped = _clip_line_poly(gdf.explode().reset_index(level=[1]), poly)
+
+    lines = clipped[
+        (clipped.geometry.type == "MultiLineString")
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return pd.concat([poly_diss, line_diss])
+
+
+def clip(gdf, clip_obj):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS) and will
+    be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj,
+    data from gdf will be clipped to the total boundary of
+    all polygons in clip_obj.
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+          Polygon vector layer used to clip gdf.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with gdf.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from gdf clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global capital cities) with a polygon (the South American continent):
+
+        >>> import geopandas as gpd
+        >>> import geopandas.clip as gc
+        >>> world = geopandas.read_file(geopandas.datasets.get_path('naturalearth_lowres'))
```

Comment:
```suggestion
        >>> world = geopandas.read_file(
        >>>      geopandas.datasets.get_path('naturalearth_lowres')
        >>> )
```
This line is failing black in the PR